include/xtensor/xbuilder.hpp

```diff
@@ -202,6 +202,121 @@ namespace xt
         private:
             int m_k;
         };
+
+        template <class T>
+        struct diagonal_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            diagonal_fn(const T& arr) : m_arr(arr)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& /*end*/) const
+            {
+                return m_arr(*begin, *begin);
+            }
+
+        private:
+            const xtype& m_arr;
+        };
+
+        template <class T>
+        struct diag_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            diag_fn(const T& arr) : m_arr(arr)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& /*end*/) const
+            {
+                auto other = begin;
+                other += 1;
+                return *begin == *other ? m_arr(*begin) : value_type(0);
+            }
+
+        private:
+            const T& m_arr;
+        };
+
+        template <class T>
+        struct flipud_fn
+        {
+            using xtype = T;
+            using value_type = typename T::value_type;
+
+            flipud_fn(const T& arr) : m_arr(arr), shape_first(m_arr.shape().front() - 1)
+            {
+            }
+
+            template <class It>
+            inline value_type operator()(const It& begin, const It& end) const
+            {
+                xindex idx(begin, end);
+                idx.front() = shape_first - idx.front();
```

Comment:
The dynamic allocation inside ``xindex`` will be done at each element access. This is a performance killer. I think this functor should be implemented like ``fn_impl`` and use an array allocated on the stack of rimplementing ``operator()(Args...)`` then forward the right values to the ``element`` and ``operator[]`` functions.